src/algorithms/ADP.jl

```diff
@@ -65,30 +82,48 @@ function set_data!(adp::CTValueIterationADP, data_new)
 end
 
 Φ(adp::CTValueIterationADP) = (x) -> reshape(adp.V̂.basis(x), 1, :)
-Ψ(adp::CTValueIterationADP) = (x, u) -> reshape(adp.dV̂.basis(vcat(x, u)), 1, :)
-function Ĥ(adp::CTValueIterationADP, r::Function)
-    _Ψ = Ψ(adp::CTValueIterationADP)
-    return (x, u, c) -> dot(_Ψ(x, u), c) + r(x, u)
+Ψ(adp::CTValueIterationADP) = (x, u) -> reshape(adp.dV̂.basis(vcat(x, u)), 1, :)  # 1 × N_Ψ
+function Ĥ(adp::CTValueIterationADP)
+    r = adp.running_cost
+    return (x, u, c) -> dot(Ψ(adp)(x, u), c) + r(x, u)
 end
 
-function update!(adp::CTValueIterationADP, running_cost, lr;
-        u_norm_max::Real)
-    @assert u_norm_max > 0
-    @unpack m, ΣΦᵀΦ_inv, data, Θ = adp
-    _Ĥ = Ĥ(adp, running_cost)
-    _Φ = Φ(adp)
-    min_Ĥ = function(x, c)
+"""
+Minimise approximate Hamiltonian.
+"""
+function min_Ĥ(adp::CTValueIterationADP)
+    # TODO: Add inexact way of obtaining minimums;
+    # see https://github.com/JinraeKim/nonlinear-VI/blob/main/nonlinear_sys.R#L98
+    @unpack m, u_norm_max = adp
+    return function (x, c)
         opt = NLopt.Opt(:LN_COBYLA, m)
-        opt.min_objective = (u, grad) -> _Ĥ(x, u, c)
+        opt.min_objective = (u, grad) -> Ĥ(adp)(x, u, c)
         opt.xtol_rel = 1e-3
         u_norm_const = function (u, grad)
             norm(u) - u_norm_max
         end
         inequality_constraint!(opt, u_norm_const)
         (minf, minx, ret) = NLopt.optimize(opt, rand(m))
     end
+end
+
+function update!(adp::CTValueIterationADP, lr)
+    @unpack m, ΣΦᵀΦ_inv, data, Θ = adp
     xs = data.states
-    term2 = xs |> Map(x -> _Φ(x)' * min_Ĥ(x, Θ*adp.V̂.w)[1]) |> collect |> sum
-    adp.V̂.w += lr * ΣΦᵀΦ_inv * term2
+    term2 = xs |> Map(x -> Φ(adp)(x)' * min_Ĥ(adp)(x, Θ*adp.V̂.param)[1]) |> collect |> sum
+    adp.V̂.param += lr * ΣΦᵀΦ_inv * term2
     nothing
 end
+
+"""
+Infer the approximate optimal input.
+"""
+function approximate_optimal_input(adp::CTValueIterationADP)
+    # TODO: Add inexact way of obtaining approximate optimal input;
+    # see https://github.com/JinraeKim/nonlinear-VI/blob/main/nonlinear_sys.R#L182
+    ŵ_f = adp.V̂.param
+    ĉ_f = adp.Θ * ŵ_f
+    return function (x, p, t)
+        û = min_Ĥ(adp)(x, ĉ_f)[2]
+    end
+end
```

Comment:
Is this not used yet?

Reply:
See [here](https://github.com/JinraeKim/FlightSims.jl/pull/8/files#diff-8908bd017b8d9c0a61bbd2a586d6a1776985ccf2805d9ae6cb8084cff170b6a8R98).